src/tfr_dumping/src/dumping_action_server.cpp

```diff
@@ -0,0 +1,164 @@
+#include <dumping_action_server.h>
+
+/*
+ * The dumping action server, it backs up the rover into the navigational aid
+ * slowly.
+ *
+ * Its first step is to make sure it can see the aruco board, it will abort the
+ * mission if it can't. 
+ *
+ * I backs up at a set speed until it get's really close and loses sight of the
+ * board. When it is blind, it drives straight back, goes slower. 
+ *
+ * It stops when the light detector get's triggered.
+ *
+ * It requires a service where it can get the most recent image on demand for
+ * the camera of interest for backing up. 
+ * */
+Dumper::Dumper(ros::NodeHandle &node, const std::string &service_name,
+        const DumpingConstraints &c) :
+    server{node, "dump", boost::bind(&Dumper::dump, this, _1), false},
+    detector{"light_detection"},
+    //TODO add dumping controller here
+    //TODO add aruco here
+    constraints{c}
+{
+    ROS_INFO("dumping action server initializing");
+    detector.waitForServer();
+
+    image_client = node.serviceClient<tfr_msgs::WrappedImage>(service_name);
+
+    velocity_publisher = node.advertise<geometry_msgs::Twist>("cmd_vel", 10);
+
+    server.start();
+    ROS_INFO("dumping action server initialized");
+}
+
+/*
+ * The business logic of the action server.
+ */
+void Dumper::dump(const tfr_msgs::EmptyGoalConstPtr &goal) 
+{  
+    //check to make sure we can see the board
+    auto initial_estimate = get_aruco_estimate();
+    if (initial_estimate.markers_detected == 0)
+    {
+        server.setAborted();
+        return;
+    }
```

Comment:
`dump` is being called by Executive, right?  Do we have a plan in place for what to do if we can't see the board here?

It might be good to use a non-empty goal that has some error codes as a return value so that Exec can attempt to take some appropriate action.  Alternatively, I can see an argument that says the Dumping system needs to try and handle this problem somehow.

Reply:
You know we haven't talked about it. From what I can think of there are 3 scenarios.

1. We are pointing in the wrong direction and have room to turn: If this is the case navigation did something really wrong, and we need to run localization again. Since executive ran localization the first time, I think executive should do it again.

2. We are pointing the right direction and we are too close to see the marker: then we should not try and localize, because we will hit the wall and just back up blind. 

3. We are pointing the wrong direction and too close: at this point we are just pretty screwed, we can't turn effectively, and don't know where we are, we should have manual control take over.

It's kind of hard to write code to cover all 3 scenarios well, if we end up any of them our positional information might be compromised or untrustworthy, so it would be difficult to determine a clean solution with our current sensor sources. 

What I suggest we can do is make one of the scenarios more likely than the other, depending on where we set the finish line, and optimize for one or the other using the setup we have.

I think it's a little less risky from a collision standpoint to try and run dumping as far away from the obstacle zone. The way we have set up dumping and planned it isn't obstacle aware, and might take a really long time to make it obstacle aware.

If we choose to use fiducial odom publisher we could get some pretty good localization data when we are close we have some more tools at our disposal when we can see the bin. We also have another one of those hd  robotics cams, so that would be pretty easy to incorporate that forward facing.

Regardless, Let's talk about it at the meeting and make it next pull request.

Reply:
All three of those scenarios seem pretty unlikely.  If navigation fails enough that we're in this situation, it may be best to switch to tele-op.  I'm not sure.

Reply:
@RyanBerge Hey we forgot to talk about this at the meeting. 

Here is what I propose. Let's make a maintenence comment here to address this in testing. However I don't think we know enough to anything here yet.